### DISCUSSION TITLE
Combining SSR/SSG with swr to fetch on client-side

### DISCUSSION THREAD

**Discussion by ben-me** [neuefische/questions#199] [Web]
Hello,

my page for dynamic routes "/[roomId].js" uses SSR to hydrate the site on the server before giving it to the client.
As far as I can tell from the documentation, the server-functions "getStaticPaths" and "getStaticProps" are running on the backend/server and only do so once when the page is requested.
The data I am fetching is currently stored in a "roomServices.js" in a "services" folder and not yet in a Database (will be added later).

Code for this looks like this:
```jsx

export async function getStaticPaths() {
  const allRooms = await getAllRooms();
  const paths = allRooms.map((rooms) => {
    return { params: { roomId: rooms.id } };
  });

  return {
    paths,
    fallback: false,
  };
}

export async function getStaticProps(context) {
  const roomId = context.params.roomId;
  const roomDetails = await getRoomById(roomId);

  return {
    props: { roomDetails },
  };
}

export default function RoomPage({ roomDetails }) {
...
}
```

I am currently trying to implement the functionality for the user to add a new reservation to the locally saved data. Since the data only gets generated once on the server, the new updated data will not be refetched and in turn not be displayed dynamically.
The nextJS Doc is recommending SWR to implement additional client side fetching.

Since most examples work with api urls and JSON Data, i've tried to adapt it to my logic (getting data from a local .js file) but had no success. 
Here is what I tried
``` jsx
// declared fetcher outside of all functions
const fetcher = (id) => getRoomById(id);

// declaring useSWR Hook with key "roomDetails.id" and the fetcher
export default function RoomPage({ roomDetails }) {
  const { data } = useSWR(roomDetails.id, fetcher);

//placing SWR config around all the rendered elements
<SWRConfig value={{ fallback: roomDetails }}>

```

I'm logging the 'data'-element of the SWR hook. At first it's undefinded (because it only gets values after hydration) but then gets filled with the roomDetails from getStaticProps.
How do I have to use the SWR correctly to implement client side fetching? I'm not really sure if I even understand the documentation and don't know where the error is.

The goal is for the user to create a new reservation and immediatly render the newly saved entry.
Here is my (unfinished) pull request:
https://github.com/ben-me/capstone-project/pull/20

**Comment by NikRoe** [comment]
Hey Benedikt,
can you send us a link to a Pull Request?

Thanks in advance

**Reply by ben-me** [reply]
Hey Niklas,

sure, here you go:
https://github.com/ben-me/capstone-project/pull/20

**Reply by NikRoe** [reply]
We solved the issue by making the `data` useSWR fetches for us optional by using [optional chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining). The code now looks approximately like this:

```js
        //...
        <Header title={data?.name} />
        <Calendar onChangeDate={changeDate} stateDate={selectedDate} />
        <DeskList>
          {data?.desks.map((desk) => (
            <DeskItem
              key={desk.id}
              deskDetails={desk}
              onShowDetails={showDetails}
              currentHighlightedDesk={highlightedDesk?.id}
            />
          ))}
        </DeskList>
        //...
```

**Comment by pixelass** [comment]
You pass the `roomDetails` into the props of your component/page, so using **SWR is not needed here**:

(See comments in code)
```jsx

const fetcher = (id) => getRoomById(id); // gets the room by id

export default function RoomPage({ roomDetails }) { // roomDetails is already the entire data of that room
  const { data } = useSWR(roomDetails.id, fetcher); // you use the id to get data from SWR
// Remember, you don't need SWR here, because roomDetails is the same as data
```

Because: here you already do the same thing as in SWR: (See comment in code)

```jsx
export async function getStaticProps(context) {
  const roomId = context.params.roomId;
  const roomDetails = await getRoomById(roomId); // gets the room by id

  return {
    props: { roomDetails }, // This is passed into the component, this is the entire room
  };
}
```


---

EDIT

This answers the question for now. Since you don't fetch anything yet, you don't need to implement the mechanism. SWR is only needed if the data is modified at some point. When you get to that point you can look at [this example from the SWR documentation](https://swr.vercel.app/examples/ssr) or at [this one here](https://swr.vercel.app/docs/with-nextjs#pre-rendering-with-default-data). Remember, even when fetching from an API, you only need SWR if you edit the data and want to keep the client in sync

You are trying to use SWR to modify local data. This is not how SWR is meant to be used. This would be done with state in react. 

* State => local Data
* Fetching & SSG/SSR => remote data (API)

**Reply by ben-me** [reply]
Hey Gregor,

thanks for your answer and the additional information.
You are right that the data I'm modifying right now is only saved locally and changes will not persist when reloading the app. Later on I'll switch from a local file to a mongoDB with api routes which will allow me to CRUD the data.
Although I already have the props for SSR from getStaticProps, the rendered data for a reservation should be updated on client side as soon as I create a new entry. Using only the SSR props, this was (as far as I understand) not possible.

Right now SWR doesn't make sense in my use case but its done with the (later) database in mind. So that I don't have to rewrite the useStates and everything when switching.

We managed to make it work with my local data by using operational chaining for the affected variables.
